lightning/src/ln/channelmanager.rs

```diff
@@ -4048,6 +4166,114 @@ where
+					ChannelPhase::UnfundedOutboundV2(chan) => {
+						chan.begin_interactive_funding_tx_construction(&self.signer_provider,
+							&self.entropy_source, self.get_our_node_id(), funding_inputs)?
+					},
+					ChannelPhase::UnfundedInboundV2(chan) => {
+						chan.begin_interactive_funding_tx_construction(&self.signer_provider,
+							&self.entropy_source, self.get_our_node_id(), funding_inputs)?
+					},
+					_ => {
+						return Err(APIError::ChannelUnavailable {
+							err: format!("Channel with ID {} is not an unfunded V2 channel", counterparty_node_id) });
+					},
+				};
+				if let Some(tx_msg) = tx_msg_opt {
+					let msg_send_event = match tx_msg {
+						InteractiveTxMessageSend::TxAddInput(msg) => events::MessageSendEvent::SendTxAddInput {
+							node_id: *counterparty_node_id, msg },
+						InteractiveTxMessageSend::TxAddOutput(msg) => events::MessageSendEvent::SendTxAddOutput {
+							node_id: *counterparty_node_id, msg },
+						InteractiveTxMessageSend::TxComplete(msg) => events::MessageSendEvent::SendTxComplete {
+							node_id: *counterparty_node_id, msg },
+					};
+					peer_state.pending_msg_events.push(msg_send_event);
+				}
+			},
+			hash_map::Entry::Vacant(_) => return Err(APIError::ChannelUnavailable {err: format!(
+				"Channel with id {} not found for the passed counterparty node_id {}",
+				channel_id, counterparty_node_id) }),
+		}
+		Ok(())
+	}
+
+	/// Handles a signed funding transaction generated by interactive transaction construction and
+	/// provided by the client.
+	///
+	/// Do NOT broadcast the funding transaction yourself. When we have safely received our
+	/// counterparty's signature(s) the funding transaction will automatically be broadcast via the
+	/// [`BroadcasterInterface`] provided when this `ChannelManager` was constructed.
+	#[cfg(dual_funding)]
+	pub fn funding_transaction_signed(&self, channel_id: &ChannelId, counterparty_node_id: &PublicKey,
+		transaction: Transaction) -> Result<(), APIError> {
+		let witnesses: Vec<_> = transaction.input.into_iter().enumerate().filter_map(|(idx, input)| {
+			if input.witness.is_empty() { None } else { Some(input.witness) }
+		}).collect();
+
+		let per_peer_state = self.per_peer_state.read().unwrap();
+		let peer_state_mutex = per_peer_state.get(counterparty_node_id)
+			.ok_or_else(|| APIError::ChannelUnavailable { err: format!("Can't find a peer matching the passed counterparty node_id {}", counterparty_node_id) })?;
+
+		let mut peer_state_lock = peer_state_mutex.lock().unwrap();
+		let peer_state = &mut *peer_state_lock;
+
+		match peer_state.channel_by_id.get_mut(channel_id) {
+			Some(ChannelPhase::Funded(chan)) => {
+				chan.verify_interactive_tx_signatures(&witnesses);
+				if let Some(ref mut signing_session) = chan.interactive_tx_signing_session {
+					if let Some(tx_signatures) = signing_session.provide_holder_witnesses(*channel_id, witnesses) {
+						peer_state.pending_msg_events.push(events::MessageSendEvent::SendTxSignatures {
```

Comment:
I suspect the msg flow is not fully correct here; we may send out tx_signatures after tx finalization and our sent commitment_signed, but before receiving commitment_signed. Tx_signatures should be sent out only after receiving commitment_signed.

Reply:
`provide_holder_witnesses()` will only return `Some` if we have already received a commitment_signed, otherwise it's `None`.

Reply:
Correct, I was wrong.
I also realized that the funding transaction can get complete right after the local signature, if that comes after a received `tx_signatures`. Therefore, I had to extend the splicing logic to be able to save the new funding tx in that case, just after calling `provide_holder_signatures`. I have extended `provide_holder_signatures()` to also check for the availability of finalized transactions.
I am not sure if this applies to general dual funding or needed only for splicing. The change can be seen in this small commit:
https://github.com/optout21/rust-lightning/commit/ac452522ac95a0fbe6cc5ddd934d860f228e9cd4
and the preceeding ones (84d9838 46f51b4 f2ba543 2011f60)